epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt

```diff
@@ -31,130 +32,222 @@ import java.util.concurrent.Executor
  * A PagedList stream wrapper that caches models built for each item. It tracks changes in paged lists and caches
  * models for each item when they are invalidated to avoid rebuilding models for the whole list when PagedList is
  * updated.
+ *
+ * The PagedList submitted to this cache must be kept in sync with the model cache. To do this,
+ * the executor of the PagedList differ is set to the same thread as the model building handler.
+ * However, change notifications from the PageList happen on that list's notify executor which is
+ * out of our control, and we require the user to configure that properly, or an error is thrown.
+ *
+ * There are two special cases:
+ *
+ * 1. The first time models are built happens synchronously for immediate UI. In this case we don't
+ * use the model cache (to avoid data synchronization issues), but attempt to fill the cache with
+ * the models later.
+ *
+ * 2. When a list is submitted it can trigger update callbacks synchronously. Since we don't control
+ * that thread we allow a special case of cache modification when a new list is being submitted,
+ * and all cache access is marked with @Synchronize to ensure safety when this happens.
  */
 internal class PagedListModelCache<T>(
     private val modelBuilder: (itemIndex: Int, item: T?) -> EpoxyModel<*>,
     private val rebuildCallback: () -> Unit,
-    private val itemDiffCallback : DiffUtil.ItemCallback<T>,
-    private val diffExecutor : Executor? = null,
-    private val modelBuildingHandler : Handler
+    itemDiffCallback: DiffUtil.ItemCallback<T>,
+    private val diffExecutor: Executor? = null,
+    private val modelBuildingHandler: Handler
 ) {
-  /**
-   * Backing list for built models. This is a full array list that has null items for not yet build models.
-   */
-  private val modelCache = arrayListOf<EpoxyModel<*>?>()
-  /**
-   * Tracks the last accessed position so that we can report it back to the paged list when models are built.
-   */
-  private var lastPosition: Int? = null
-
-  /**
-   * Observer for the PagedList changes that invalidates the model cache when data is updated.
-   */
-  private val updateCallback = object : ListUpdateCallback {
-    override fun onChanged(position: Int, count: Int, payload: Any?) {
-      (position until (position + count)).forEach {
-        modelCache[it] = null
-      }
-      rebuildCallback()
-    }
+    /**
+     * Backing list for built models. This is a full array list that has null items for not yet build models.
+     */
+    private val modelCache = arrayListOf<EpoxyModel<*>?>()
+    /**
+     * Tracks the last accessed position so that we can report it back to the paged list when models are built.
+     */
+    private var lastPosition: Int? = null
+
+    /**
+     * Set to true while a new list is being submitted, so that we can ignore the update callback
+     * thread restriction.
+     */
+    private var inSubmitList: Boolean = false
```

Comment:
there is a race condition where submit might be called in thread A at the same time we are checking this value. Needs to be thread local

Reply:
I think the `@Synchronized` prevent that from happening, since all access to this variable is behind synchronized functions